packages/dds/tree/src/changeset/markListFactory.ts

```diff
@@ -0,0 +1,140 @@
+        for (const item of offsetOrContent) {
+            if (typeof item === "number") {
+                this.pushOffset(item);
+            } else {
+                this.pushContent(item);
+            }
+        }
+    }
+
+    public pushOffset(offset: number): void {
+        this.offset += offset;
+    }
+
+    public pushContent(mark: T.ObjectMark | T.AttachGroup): void {
+        if (this.offset > 0) {
+            this.list.push(this.offset);
+            this.offset = 0;
+        }
+        const prev = this.list[this.list.length - 1];
+        if (this.offset === 0 && prev !== undefined && typeof prev === "object") {
+            if (isAttachGroup(prev)) {
+                if (isAttachGroup(mark)) {
+                    const lastLeft = prev[prev.length - 1];
+                    const firstRight = mark[0];
+                    if (
+                        lastLeft !== undefined
+                        && firstRight !== undefined
+                        && lastLeft.type === firstRight.type
+                        && lastLeft.id === firstRight.id
+                        && lastLeft.id === firstRight.id
+                    ) {
+                        const type = lastLeft.type;
+                        switch (type) {
+                            case "Insert":
+                            case "MoveIn": {
+                                const firstRightAttach = firstRight as T.Insert | T.MoveIn;
+                                if (
+                                    lastLeft.heed === firstRightAttach.heed
+                                    && lastLeft.tiebreak === firstRightAttach.tiebreak
+                                    && lastLeft.src?.id === firstRightAttach.src?.id
+                                    && lastLeft.src?.seq === firstRightAttach.src?.seq
+                                    && lastLeft.scorch?.id === firstRightAttach.scorch?.id
+                                    && lastLeft.scorch?.seq === firstRightAttach.scorch?.seq
+                                ) {
+                                    if (lastLeft.type === "Insert") {
+                                        const firstRightInsert = firstRight as T.Insert;
+                                        lastLeft.content.push(...firstRightInsert.content);
+                                    } else {
+                                        const firstRightMoveIn = firstRight as T.MoveIn;
+                                        lastLeft.count += firstRightMoveIn.count;
+                                    }
+                                    prev.push(...mark.slice(1));
+                                    return;
+                                }
+                                break;
+                            }
+                            default: break;
+                        }
+                    }
+                    prev.push(...mark);
+                    return;
+                }
```

Comment:
I went for `extendAttachGroup` since it has an API where one of the parameters gets mutated.
The awkward API contract is part of why I didn't make a helper. Let me know if you think it's worth the perf hit to produce a whole new attach group without mutating either inputs.